iroh/src/magicsock.rs

```diff
@@ -3022,28 +3073,29 @@ fn new_re_stun_timer(initial_delay: bool) -> time::Interval {
     }
 }
 
-/// Initial connection setup.
 #[cfg(not(wasm_browser))]
-fn bind(
-    addr_v4: Option<SocketAddrV4>,
-    addr_v6: Option<SocketAddrV6>,
-) -> Result<(UdpConn, Option<UdpConn>)> {
-    let addr_v4 = addr_v4.unwrap_or_else(|| SocketAddrV4::new(Ipv4Addr::UNSPECIFIED, 0));
-    let pconn4 = UdpConn::bind(SocketAddr::V4(addr_v4)).context("bind IPv4 failed")?;
-
-    let ip4_port = pconn4.local_addr()?.port();
-    let ip6_port = ip4_port.checked_add(1).unwrap_or(ip4_port - 1);
-    let addr_v6 =
-        addr_v6.unwrap_or_else(|| SocketAddrV6::new(Ipv6Addr::UNSPECIFIED, ip6_port, 0, 0));
-    let pconn6 = match UdpConn::bind(SocketAddr::V6(addr_v6)) {
-        Ok(conn) => Some(conn),
+fn bind_with_fallback(mut addr: SocketAddr) -> anyhow::Result<UdpSocket> {
+    debug!(%addr, "binding");
+
+    // First try binding a preferred port, if specified
+    match UdpSocket::bind_full(addr) {
+        Ok(socket) => {
+            let local_addr = socket.local_addr().context("UDP socket not bound")?;
+            debug!(%addr, %local_addr, "successfully bound");
+            return Ok(socket);
+        }
         Err(err) => {
-            info!("bind ignoring IPv6 bind failure: {:?}", err);
-            None
+            debug!(%addr, "failed to bind: {err:#}");
+            // If that was already the fallback port, then error out
+            if addr.port() == 0 {
+                return Err(err.into());
+            }
         }
-    };
+    }
 
-    Ok((pconn4, pconn6))
+    // Otherwise, try binding with port 0
+    addr.set_port(0);
+    Ok(UdpSocket::bind_full(addr).context("failed to bind on fallback port")?)
```

Comment:
The diff here is a little rough and confusing.

The old `fn bind` still exists: It moved to `ActorSocketState::bind`.
The new `fn bind_with_fallback` is a cleaned-up version of what used to be `udp_conn::bind`/`UdpConn::bind`.